### PR TITLE
Allow fine-grained control over rgahp SSH BatchMode (SOFTWARE-4239)

### PR DESCRIPTION
**Pull request by brianhlin**
When connecting to some HPCs with "-o BatchMode yes", we get kicked to
a password or PIN prompt. This gives administrators an option of
disabling BatchMode

https://opensciencegrid.atlassian.net/browse/SOFTWARE-4239

### CODE REVIEW

File: src/condor_gridmanager/remote_gahp [hunk 1]
```diff
@@ -138,10 +143,14 @@ fi
...
-#echo "** Follows output of: ssh -o \"BatchMode yes\" $REMOTE_USER@$REMOTE_HOSTNAME /bin/bash -c \"'GLITE_LOCATION=$REMOTE_GLITE $REMOTE_GLITE/bin/batch_gahp $*'\""
+SSH_ARGS=()
+# Some HPCs require a PIN when specifying "-o BatchMode yes", so we should allow users
+# https://opensciencegrid.atlassian.net/browse/SOFTWARE-4239
+[[ $SSH_BATCHMODE == "yes" ]] && SSH_ARGS+=(-o "BatchMode yes")
+
+#echo "** Follows output of: ssh ${SSH_ARGS[@]} $REMOTE_USER@$REMOTE_HOSTNAME /bin/bash -c \"'GLITE_LOCATION=$REMOTE_GLITE $REMOTE_GLITE/bin/batch_gahp $*'\""
 if [ "${REMOTE_CMD}" = "batch_gahp" ] ; then
-    ssh -o "BatchMode yes" $REMOTE_USER@$REMOTE_HOSTNAME /bin/bash -l -c "'GLITE_LOCATION=$REMOTE_GLITE $REMOTE_GLITE/bin/batch_gahp $*'"
+    ssh ${SSH_ARGS[@]} $REMOTE_USER@$REMOTE_HOSTNAME /bin/bash -l -c "'GLITE_LOCATION=$REMOTE_GLITE $REMOTE_GLITE/bin/batch_gahp $*'"
```
Comment (edquist) [inline src/condor_gridmanager/remote_gahp:153]:
```suggestion
    ssh "${SSH_ARGS[@]}" $REMOTE_USER@$REMOTE_HOSTNAME /bin/bash -l -c "'GLITE_LOCATION=$REMOTE_GLITE $REMOTE_GLITE/bin/batch_gahp $*'"
```

You need to quote an expanded array if you want whitespace to be preserved in the arguments.  And, in this case you do for `-o "BatchMode yes"`

File: src/condor_gridmanager/remote_gahp [hunk 2]
```diff
@@ -160,14 +169,14 @@ elif [ "${REMOTE_CMD}" = "condor_ft-gahp" ] ; then
...
-        SSH_ARGS="-o ExitOnForwardFailure=yes"
+        SSH_ARGS+=(-o ExitOnForwardFailure=yes)
         tries=3
     else
         tries=1
     fi
     while ((tries-- > 0 && SSH_STATUS == 255)) ; do
         let port=${RANDOM}+32768
-        ssh $SSH_ARGS -R $port:$GRIDMANAGER_ADDRESS -o "BatchMode yes" $REMOTE_USER@$REMOTE_HOSTNAME /bin/bash -l -c "'echo Allocated port $port for remote forward to 1>&2 ; CONDOR_CONFIG=$REMOTE_GLITE/etc/condor_config.ft-gahp $REMOTE_GLITE/bin/condor_ft-gahp -f $*'"
+        ssh ${SSH_ARGS[@]} -R $port:$GRIDMANAGER_ADDRESS $REMOTE_USER@$REMOTE_HOSTNAME /bin/bash -l -c "'echo Allocated port $port for remote forward to 1>&2 ; CONDOR_CONFIG=$REMOTE_GLITE/etc/condor_config.ft-gahp $REMOTE_GLITE/bin/condor_ft-gahp -f $*'"
```
Comment (edquist) [inline src/condor_gridmanager/remote_gahp:179]:
```suggestion
        ssh "${SSH_ARGS[@]}" -R $port:$GRIDMANAGER_ADDRESS $REMOTE_USER@$REMOTE_HOSTNAME /bin/bash -l -c "'echo Allocated port $port for remote forward to 1>&2 ; CONDOR_CONFIG=$REMOTE_GLITE/etc/condor_config.ft-gahp $REMOTE_GLITE/bin/condor_ft-gahp -f $*'"
```

File: src/condor_gridmanager/remote_gahp [hunk 3]
```diff
@@ -44,6 +46,7 @@ Options: \n \
  --rgahp-nopass \t\tno passphrase protecting ssh private key (same as empty rgahp-pass)\n \
  --rgahp-glite PATH \tpath to the directory of the script (~/bosco/glite)\n \
  --rgahp-script SCRIPT \tpath to script to start up blahp (PATH/bin/batch_gahp)\n \
+ --rgahp-nobatchmode \t\tDisable SSH BatchMode (yes)\n \
```
Comment (edquist) [inline src/condor_gridmanager/remote_gahp:49]: Unrelated to your change, but -- wow, this usage string is something else.

**Review by edquist — APPROVED** [review]
pre approving, but please apply the quoting fixes

File: src/condor_gridmanager/remote_gahp (hunk 3 above)
Comment (brianhlin) [reply src/condor_gridmanager/remote_gahp]: Yeah, I was just blindly copying previous usage strings but it doesn't make sense to tell the user what the default is for a flag. Hopefully it's more sensible now.

**Review by JaimeFrey — APPROVED** [review]

**Pull request merged**